src/include/main/query_result.h

```diff
@@ -10,78 +10,62 @@ using namespace kuzu::processor;
         querySummary = make_unique<QuerySummary>();
         querySummary->setPreparedSummary(preparedSummary);
     }
 
     inline bool isSuccess() const { return success; }
     inline string getErrorMessage() const { return errMsg; }
 
-    inline void setResultHeaderAndTable(std::unique_ptr<QueryResultHeader> header,
-        std::shared_ptr<processor::FactorizedTable> factorizedTable) {
-        this->header = move(header);
-        this->factorizedTable = move(factorizedTable);
-        resetIterator();
+    inline size_t getNumColumns() const { return columnDataTypes.size(); }
+    inline vector<std::string> getColumnNames() { return columnNames; }
+    inline std::vector<common::DataType> getColumnDataTypes() { return columnDataTypes; }
+
+    inline uint64_t getNumTuples() {
+        return querySummary->getIsExplain() ? 0 : factorizedTable->getTotalNumFlatTuples();
     }
 
-    bool hasNext();
+    inline QuerySummary* getQuerySummary() const { return querySummary.get(); }
 
-    // TODO: this is not efficient and should be replaced by iterator
-    std::shared_ptr<processor::FlatTuple> getNext();
+    void initResultTableAndIterator(std::shared_ptr<processor::FactorizedTable> factorizedTable_,
+        const expression_vector& columns,
+        const vector<expression_vector>& expressionToCollectPerColumn);
 
-    void writeToCSV(
-        string fileName, char delimiter = ',', char escapeCharacter = '"', char newline = '\n');
+    bool hasNext();
 
-    inline uint64_t getNumColumns() const { return header->columnDataTypes.size(); }
+    std::shared_ptr<processor::FlatTuple> getNext();
 
-    inline QuerySummary* getQuerySummary() const { return querySummary.get(); }
+    void writeToCSV(const string& fileName, char delimiter = ',', char escapeCharacter = '"',
+        char newline = '\n');
 
     // TODO: interfaces below should be removed
     // used in shell to walk the result twice (first time getting maximum column width)
-    inline void resetIterator() {
-        iterator = make_unique<FlatTupleIterator>(*factorizedTable, header->columnDataTypes);
-    }
-
-    inline uint64_t getNumTuples() {
-        return querySummary->getIsExplain() ? 0 : factorizedTable->getTotalNumFlatTuples();
-    }
-
-    inline vector<std::string> getColumnNames() { return header->columnNames; }
-
-    inline std::vector<common::DataType> getColumnDataTypes() { return header->columnDataTypes; }
+    inline void resetIterator() { iterator->resetState(); }
 
 private:
     void validateQuerySucceed();
 
+private:
+    // execution status
     bool success = true;
     std::string errMsg;
 
-    std::unique_ptr<QueryResultHeader> header;
+    // header information
+    std::vector<std::string> columnNames;
+    std::vector<common::DataType> columnDataTypes;
+    // data
     std::shared_ptr<processor::FactorizedTable> factorizedTable;
     std::unique_ptr<processor::FlatTupleIterator> iterator;
+    shared_ptr<FlatTuple> tuple;
```

Comment:
1. There is an inconsistency here (shared_ptr) and in the `initResultTableAndIterator` function (make_unique).
2. Can we not use a `unique_ptr` here? And we return a `FlatTuple*` for the `getNext` function.

Reply:
Using unique_ptr meaning we will need to expose getCopy and getReference interfaces. Ducked also uses a shared_ptr. So I wouldn't priority changing this.